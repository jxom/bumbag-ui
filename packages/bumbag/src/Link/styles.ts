import { cssClass } from '../styled';
import { palette, theme, fontWeight } from '../utils';

export const Link = (styleProps) => cssClass`
  color: ${palette('primary', { dark: 'primary300' })(styleProps)};
  fill: ${palette('primary', { dark: 'primary300' })(styleProps)};
  cursor: pointer;
  font-weight: ${fontWeight('semibold')(styleProps)};
  text-decoration: none;

  &:hover {
    text-decoration: underline;

    & {
      ${theme(styleProps.themeKey, `styles.hover`)(styleProps)};
    }
  }

  &:focus {
    & {
      ${theme(styleProps.themeKey, `styles.focus`)(styleProps)};
    }
  }

  & .bb-Icon {
    vertical-align: -0.125em;
  }

  & {
    ${theme(styleProps.themeKey, `styles.base`)(styleProps)};
  }
`;
