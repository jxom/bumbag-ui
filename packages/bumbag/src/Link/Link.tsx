import { Box as ReakitBox } from 'reakit';

import { useClassName, createComponent, createElement, createHook } from '../utils';
import { Box, BoxProps } from '../Box';

import * as styles from './Link.styles';

export type LocalLinkProps = {};
export type LinkProps = BoxProps & React.AnchorHTMLAttributes<any> & LocalLinkProps;

const useProps = createHook<LinkProps>(
  (props, { themeKey }) => {
    const boxProps = Box.useProps(props);

    const className = useClassName({
      style: styles.Link,
      styleProps: props,
      themeKey,
      prevClassName: boxProps.className,
    });

    return { ...boxProps, className };
  },
  { themeKey: 'Link' }
);

export const Link = createComponent<LinkProps>(
  (props) => {
    const linkProps = useProps(props);
    return createElement({ children: props.children, component: ReakitBox, use: props.use, htmlProps: linkProps });
  },
  {
    attach: {
      useProps,
      displayName: 'Link',
    },
    defaultProps: {
      use: 'a',
    },
    themeKey: 'Link',
  }
);
