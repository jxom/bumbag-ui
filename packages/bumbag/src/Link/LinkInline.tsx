import { Box as ReakitBox } from 'reakit';

import { useClassName, createComponent, createElement, createHook } from '../utils';
import { Box, BoxProps } from '../Box';

import * as styles from './styles';

export type LocalLinkInlineProps = {};
export type LinkInlineProps = BoxProps & React.LinkHTMLAttributes<any> & LocalLinkInlineProps;

const useProps = createHook<LinkInlineProps>(
  (props, { themeKey, themeKeyOverride }) => {
    const boxProps = Box.useProps(props);

    const className = useClassName({
      style: styles.LinkInline,
      styleProps: props,
      themeKey,
      themeKeyOverride,
      prevClassName: boxProps.className,
    });

    return { ...boxProps, className };
  },
  { themeKey: 'Link.Inline' }
);

export const LinkInline = createComponent<LinkInlineProps>(
  (props) => {
    const linkInlineProps = useProps(props);
    return createElement({
      children: props.children,
      component: ReakitBox,
      use: props.use,
      htmlProps: linkInlineProps,
    });
  },
  {
    attach: {
      useProps,
      displayName: 'Link.Inline',
    },
    defaultProps: {
      use: 'a',
    },
    themeKey: 'Link.Inline',
  }
);
