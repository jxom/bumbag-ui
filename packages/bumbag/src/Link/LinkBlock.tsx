import { Box as ReakitBox } from 'reakit';

import { useClassName, createComponent, createElement, createHook } from '../utils';
import { Box, BoxProps } from '../Box';

import * as styles from './Link.styles';

export type LocalLinkBlockProps = {};
export type LinkBlockProps = BoxProps & React.AnchorHTMLAttributes<any> & LocalLinkBlockProps;

const useProps = createHook<LinkBlockProps>(
  (props, { themeKey }) => {
    const boxProps = Box.useProps(props);

    const className = useClassName({
      style: styles.LinkBlock,
      styleProps: props,
      themeKey,
      prevClassName: boxProps.className,
    });

    return { ...boxProps, className };
  },
  { themeKey: 'Link.Block' }
);

export const LinkBlock = createComponent<LinkBlockProps>(
  (props) => {
    const linkBlockProps = useProps(props);
    return createElement({ children: props.children, component: ReakitBox, use: props.use, htmlProps: linkBlockProps });
  },
  {
    attach: {
      useProps,
      displayName: 'Link.Block',
    },
    defaultProps: {
      use: 'a',
    },
    themeKey: 'Link.Block',
  }
);
