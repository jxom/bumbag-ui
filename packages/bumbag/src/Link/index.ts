import * as styles from './styles';
import { Link as _Link } from './Link';
import { LinkBlock } from './LinkBlock';
import { LinkInline } from './LinkInline';

export * from './Link';
export const Link = Object.assign(_Link, {
  Block: LinkBlock,
  Inline: LinkInline,
});
export { styles as linkStyles };
