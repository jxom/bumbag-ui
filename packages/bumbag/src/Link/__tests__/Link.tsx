import * as React from 'react';
import { renderHook } from '@testing-library/react-hooks';
import { Link } from '../Link';
import render from '../../utils/_tests/render';

describe('props', () => {
  it('should render correctly', () => {
    const { container } = render(<Link>Hello world</Link>);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render correctly with CSS props', () => {
    const { container } = render(<Link color="primary">Hello world</Link>);
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe('composition', () => {
  describe('as', () => {
    it('should render correctly', () => {
      const { container } = render(<Link use="p">Hello world</Link>);
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('hook', () => {
    it('should return with Link props', () => {
      const { result } = renderHook(() => Link.useProps());
      expect(result.current).toMatchSnapshot();
    });
  });

  describe('render props', () => {
    it('should render correctly', () => {
      const { container } = render(<Link>{(LinkProps) => <div {...LinkProps}>Hello world</div>}</Link>);
      expect(container.firstChild).toMatchSnapshot();
    });
  });
});

describe('overrides', () => {
  it('Link.base should render correctly', () => {
    const { container } = render(
      <Link overrides={{ Link: { styles: { base: { backgroundColor: 'red' } } } }}>hello world</Link>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('Link.hover should render correctly', () => {
    const { container } = render(
      <Link overrides={{ Link: { styles: { hover: { backgroundColor: 'red' } } } }}>hello world</Link>
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe('theming', () => {
  it('Link.base should render correctly', () => {
    const { container } = render(<Link>hello world</Link>, {
      theme: { Link: { styles: { base: { backgroundColor: 'red' } } } },
    });
    expect(container.firstChild).toMatchSnapshot();
  });

  it('Link.hover should render correctly', () => {
    const { container } = render(<Link>hello world</Link>, {
      theme: { Link: { styles: { hover: { backgroundColor: 'red' } } } },
    });
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe('defaultProps', () => {
  it('should render correctly for className', () => {
    const { container } = render(<Link>hello world</Link>, {
      theme: { Link: { defaultProps: { className: 'test' } } },
    });
    expect(container.firstChild).toMatchSnapshot();
  });
});
