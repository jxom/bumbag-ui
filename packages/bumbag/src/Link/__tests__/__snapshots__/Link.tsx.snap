// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`composition as should render correctly 1`] = `
.emotion-1 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-1:focus:not(:focus-visible) {
  outline: none;
}

.emotion-0 {
  color: #574feb;
  fill: #574feb;
  cursor: pointer;
  font-weight: 600;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.emotion-0:hover {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.emotion-0 .fp-Icon {
  vertical-align: -0.125em;
}

<p
  class="emotion-0 emotion-1 emotion-2 fp-Box fp-Link"
>
  Hello world
</p>
`;

exports[`composition hook should return with Link props 1`] = `
Object {
  "className": "css-a2rsep css-4lfl7m css-1yuzjtp fp-Box fp-Link",
}
`;

exports[`composition render props should render correctly 1`] = `
.emotion-1 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-1:focus:not(:focus-visible) {
  outline: none;
}

.emotion-0 {
  color: #574feb;
  fill: #574feb;
  cursor: pointer;
  font-weight: 600;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.emotion-0:hover {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.emotion-0 .fp-Icon {
  vertical-align: -0.125em;
}

<div
  class="emotion-0 emotion-1 emotion-2 fp-Box fp-Link"
>
  Hello world
</div>
`;

exports[`defaultProps should render correctly for className 1`] = `
.emotion-1 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-1:focus:not(:focus-visible) {
  outline: none;
}

.emotion-0 {
  color: #574feb;
  fill: #574feb;
  cursor: pointer;
  font-weight: 600;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.emotion-0:hover {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.emotion-0 .fp-Icon {
  vertical-align: -0.125em;
}

<a
  class="emotion-0 emotion-1 emotion-2 test fp-Box fp-Link"
>
  hello world
</a>
`;

exports[`overrides Link.base should render correctly 1`] = `
.emotion-1 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-1:focus:not(:focus-visible) {
  outline: none;
}

.emotion-0 {
  color: #574feb;
  fill: #574feb;
  cursor: pointer;
  font-weight: 600;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.emotion-0:hover {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.emotion-0 .fp-Icon {
  vertical-align: -0.125em;
}

.emotion-0.emotion-0 {
  background-color: #ff0000;
}

<a
  class="emotion-0 emotion-1 emotion-2 fp-Box fp-Link"
>
  hello world
</a>
`;

exports[`overrides Link.hover should render correctly 1`] = `
.emotion-1 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-1:focus:not(:focus-visible) {
  outline: none;
}

.emotion-0 {
  color: #574feb;
  fill: #574feb;
  cursor: pointer;
  font-weight: 600;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.emotion-0:hover {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.emotion-0:hover.emotion-0:hover {
  background-color: #ff0000;
}

.emotion-0 .fp-Icon {
  vertical-align: -0.125em;
}

<a
  class="emotion-0 emotion-1 emotion-2 fp-Box fp-Link"
>
  hello world
</a>
`;

exports[`props should render correctly 1`] = `
.emotion-1 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-1:focus:not(:focus-visible) {
  outline: none;
}

.emotion-0 {
  color: #574feb;
  fill: #574feb;
  cursor: pointer;
  font-weight: 600;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.emotion-0:hover {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.emotion-0 .fp-Icon {
  vertical-align: -0.125em;
}

<a
  class="emotion-0 emotion-1 emotion-2 fp-Box fp-Link"
>
  Hello world
</a>
`;

exports[`props should render correctly with CSS props 1`] = `
.emotion-1 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-1:focus:not(:focus-visible) {
  outline: none;
}

.emotion-0 {
  color: #574feb;
  fill: #574feb;
  cursor: pointer;
  font-weight: 600;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.emotion-0:hover {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.emotion-0 .fp-Icon {
  vertical-align: -0.125em;
}

.emotion-2 {
  color: #574feb !important;
}

<a
  class="emotion-0 emotion-1 emotion-2 fp-Box fp-Link"
>
  Hello world
</a>
`;

exports[`theming Link.base should render correctly 1`] = `
.emotion-1 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-1:focus:not(:focus-visible) {
  outline: none;
}

.emotion-0 {
  color: #574feb;
  fill: #574feb;
  cursor: pointer;
  font-weight: 600;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.emotion-0:hover {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.emotion-0 .fp-Icon {
  vertical-align: -0.125em;
}

.emotion-0.emotion-0 {
  background-color: #ff0000;
}

<a
  class="emotion-0 emotion-1 emotion-2 fp-Box fp-Link"
>
  hello world
</a>
`;

exports[`theming Link.hover should render correctly 1`] = `
.emotion-1 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-1:focus:not(:focus-visible) {
  outline: none;
}

.emotion-0 {
  color: #574feb;
  fill: #574feb;
  cursor: pointer;
  font-weight: 600;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.emotion-0:hover {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.emotion-0:hover.emotion-0:hover {
  background-color: #ff0000;
}

.emotion-0 .fp-Icon {
  vertical-align: -0.125em;
}

<a
  class="emotion-0 emotion-1 emotion-2 fp-Box fp-Link"
>
  hello world
</a>
`;
